Free functions.

chriseth opened the pull request:
Experimental PR to allow functions at file-level.

Refers to https://github.com/ethereum/solidity/issues/8787

As an extension, we could also allow `using * for uint;` or `using FileName.* for uint;` to allow binding functions to types in the same way as we bind internal library functions to types.

TODO:
 - [x] some more semantic tests
 - [x] documentation
 - [x] changelog entry

Test ideas:
- Pointers to free functions (calling them, storing them in variables).
- Calling a free function from a constructor.
- Assigning a free function pointer to a storage variable in a constructor and calling it via that pointer from another function of that contract.
- Handling internal and external function pointers inside free functions.
- Trying to call a private contract or library function from a free function.
- Passing in parameters that take multiple stack slots (e.g. external function pointers).
- Using `selfdestruct` in a free function.
- Applying a modifier to a free function.
- Trying to get a selector of a free function.
- Declaring a free function without body.
- Calling an internal contract function from inside that contract when there's a free function with the same name.
- Free function with the same name as a struct or an enum.
- Free function with dots in the name.
- Free function called `pragma()`, `contract()`, `library()`, `interface()`, `struct()`, `enum()` or `import()`.
- Free function called `gas()`.
- Recursive call of a free function.
- Trying to bind a free function with `using for`.
- Triggering events from free functions.
- Using `try .. catch` or `revert()` in free functions.
- name clash between free function and function in contract
- name clash between free function and function inherited from a contract from a different file
- should free functions be able to use `.value()`?

chriseth committed:
Free functions.

chriseth commented:
If we want that, I would remove the requirement for a visibility for free functions and set it to `internal`.

hrkrshnn commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -102,6 +102,9 @@ ASTPointer<SourceUnit> Parser::parse(shared_ptr<Scanner> const& _scanner)
 				nodes.push_back(parseEnumDefinition());
 				break;
+			case Token::Function:
+				nodes.push_back(parseFunctionDefinition());
```
This will currently allow functions without body. Not sure if this is intentional. How will they get implemented?

chriseth replied:
This PR was just a test if it can work - we need to add more restrictions.

chriseth commented:
Because `annotation().scope` is only available after reference resolution, some things get a bit more difficult. I think it would be better if the parser already specifies that the function is a free function.

chriseth commented on test/libsolidity/semanticTests/freeFunctions/new_operator.sol:
```diff
@@ -0,0 +1,15 @@
+}
+
+function test() returns (uint) {
+  return (new C()).x();
```
This test is actually passing! :)

chriseth committed:
Update existing tests.

chriseth committed:
Documentation.

hrkrshnn commented on libsolidity/analysis/SyntaxChecker.cpp:
```diff
@@ -299,19 +299,41 @@ bool SyntaxChecker::visit(ContractDefinition const& _contract)
 
+void SyntaxChecker::endVisit(ContractDefinition const&)
+{
+	m_currentContractKind = {};
```
```suggestion
	m_currentContractKind = std::nullopt;
```
A bit more explicit.

hrkrshnn commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -332,17 +335,23 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 		if (_function.libraryFunction())
 			m_errorReporter.typeError(7708_error, _function.location(), "Library functions cannot be payable.");
-		if (_function.isOrdinary() && !_function.isPartOfExternalInterface())
+		else if (_function.isFree())
+			m_errorReporter.typeError(9559_error, _function.location(), "Free functions cannot be payable.");
```
Can you have this as the first condition? Would be nice to have the checks for freeness to be first.

chriseth replied:
Does it matter?

hrkrshnn commented on libsolidity/ast/Types.cpp:
```diff
@@ -3132,7 +3131,10 @@ FunctionTypePointer FunctionType::interfaceFunctionType() const
 {
 	// Note that m_declaration might also be a state variable!
 	solAssert(m_declaration, "Declaration needed to determine interface function type.");
-	bool isLibraryFunction = kind() != Kind::Event && dynamic_cast<ContractDefinition const&>(*m_declaration->scope()).isLibrary();
+	bool isLibraryFunction = false;
+	if (kind() != Kind::Event)
+		if (auto const* contract = dynamic_cast<ContractDefinition const*>(m_declaration->scope()))
+			isLibraryFunction = contract->isLibrary();
```
Can this be called by free functions?

chriseth replied:
Calling `interfaceFunctionType` on a free function does not make sense, but I think it can happen inside the type system at some point. The bool `isLibraryFunction` would be false in this case. Do you think it is a problem?

cameel commented on docs/contracts/functions.rst:
```diff
@@ -6,6 +6,34 @@
+        bool found;
+        function f(uint[] memory _arr) public {
+            // This calls the free funtion internally.
+            // The compiler will add its code the contract.
```
```suggestion
            // The compiler will add its code to the contract.
```

cameel commented on docs/structure-of-a-contract.rst:
```diff
@@ -43,19 +43,26 @@ visibility.
         }
     }
 
+    // helper function defined outside of a contract
```
```suggestion
    // Helper function defined outside of a contract
```
Other comments in this file are all capitalized.

cameel commented on libsolidity/ast/AST.h:
```diff
@@ -788,6 +789,7 @@ class FunctionDefinition: public CallableDeclaration, public StructurallyDocumen
 		StructurallyDocumented(_documentation),
 		ImplementationOptional(_body != nullptr),
 		m_stateMutability(_stateMutability),
+		m_free(_free),
```
I think this constructor should have assertions to prevent combinations like:
- constructor as a free function
- fallback or receive as a free function
- payable free function

```c++
if (isFree())
{
	solAssert(isOrdinary(), "");
	solAssert(!isPayable(), "");
}
```

chriseth replied:
Added them but removed again because they are not enforced by the parser.

cameel commented on libsolidity/ast/AST.h:
```diff
@@ -788,6 +789,7 @@ class FunctionDefinition: public CallableDeclaration, public StructurallyDocumen
 		StructurallyDocumented(_documentation),
 		ImplementationOptional(_body != nullptr),
```
Since implementation is not optional for free functions, there should be an assert against it being missing:
```c++
if (isFree())
	solAssert(_body != nullptr, "");
```

chriseth replied:
I added this, but also reverted it again, because it leads to a "expected {" error in the parser which is not really nice for the user.

cameel commented on libsolidity/ast/AST.cpp:
```diff
@@ -373,7 +373,8 @@ FunctionDefinition const& FunctionDefinition::resolveVirtual(
 
-	solAssert(!dynamic_cast<ContractDefinition const&>(*scope()).isLibrary(), "");
+	solAssert(!isFree(), "");
+	solAssert(!libraryFunction(), "");
```
I'm not sure but maybe `solAssert(isOrdinary(), "");` should also be added here? Can fallback and receive functions be virtual?

cameel commented on test/libsolidity/semanticTests/freeFunctions/easy.sol:
```diff
@@ -0,0 +1,13 @@
+  return a + b;
+}
+
+contract c {
```
```suggestion
contract C {
```

cameel commented on test/libsolidity/semanticTests/freeFunctions/libraries_from_free.sol:
```diff
@@ -0,0 +1,21 @@
+}
+// ----
+// library: L
+// f() -> 7, 8
```
Missing newline.

cameel reviewed (commented):
Not a full review yet. So far i looked through the tests, docs, and the AST changes.

And here are some test ideas for both correct and incorrect usage:
- Pointers to free functions (calling them, storing them in variables).
- Calling a free function from a constructor.
- Assigning a free function pointer to a storage variable in a constructor and calling it via that pointer from another function of that contract.
- Handling internal and external function pointers inside free functions.
- Trying to call a private contract or library function from a free function.
- Passing in parameters that take multiple stack slots (e.g. external function pointers).
- Using `selfdestruct` in a free function.
- Applying a modifier to a free function.
- Trying to get a selector of a free function.
- Declaring a free function without body.
- Calling an internal contract function from inside that contract when there's a free function with the same name.
- Free function with the same name as a struct or an enum.
- Free function with dots in the name.
- Free function called `pragma()`, `contract()`, `library()`, `interface()`, `struct()`, `enum()` or `import()`.
- Free function called `gas()`.
- Recursive call of a free function.
- Trying to bind a free function with `using for`.
- Triggering events from free functions.
- Using `try .. catch` or `revert()` in free functions.

chriseth commented:
Rebased and fixed. Thanks for the test suggestions, @cameel 
I incorporated some of the suggestions, but most of them are really not too test-worthy, I think. The code generator just works on the compilation queue and emits code for whatever needs code emitted, and it works the same whether the function is defined inside a contract or not. Since we have the same thing already for internal library functions, it does not add much value. For example, `selfdestruct` works the same way as it works in contracts and internal library functions.

chriseth commented:
Removed the assertions in the constructor of FunctionDefinition in AST.h again because the conditions are actually not enforced by the parser.

Marenz commented on libsolidity/analysis/ViewPureChecker.cpp:
```diff
@@ -128,30 +128,23 @@ class AssemblyViewPureChecker
-
 bool ViewPureChecker::visit(FunctionDefinition const& _funDef)
 {
+	if (m_checkModifiers)
```
Maybe rename this to `m_checkModifiersOnly`? Otherwise the name doesn't make it clear that we don't check functions anymore..

Marenz commented on libsolidity/ast/Types.cpp:
```diff
@@ -375,6 +375,7 @@ TypePointer Type::commonType(Type const* _a, Type const* _b)
 
 MemberList const& Type::members(ASTNode const* _currentScope) const
 {
+	// TODO _currentScope can be nullptr!
```
Will this be fixed in this PR?

chriseth replied:
This was just a note from the beginning of the implementation work. There is an assert below that takes it being nullptr into account, so I assume it is fine.

Marenz commented on docs/contracts/functions.rst:
```diff
@@ -6,6 +6,34 @@
 Functions
 *********
 
+Functions can be defined inside contracts, but they can also be defined outside contracts.
```
`Functions can be defined inside and outside of contracts.`

Marenz commented on docs/contracts/functions.rst:
```diff
@@ -6,6 +6,34 @@
+::
+
+    // SPDX-License-Identifier: GPL-3.0
+    pragma solidity >0.6.11 <0.7.0;
```
`>0.6.12` ?

Marenz commented on docs/structure-of-a-contract.rst:
```diff
@@ -43,19 +43,26 @@ visibility.
 
     // SPDX-License-Identifier: GPL-3.0
-    pragma solidity >=0.4.0 <0.7.0;
+    pragma solidity >0.6.11 <0.7.0;
```
`>0.6.12`?

Marenz reviewed (changes requested):
Overall LGTM, some small comments

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -327,17 +329,23 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 	if (_function.markedVirtual())
 	{
-		if (_function.annotation().contract->isInterface())
+		if (_function.isFree())
+			m_errorReporter.syntaxError(4493_error, _function.location(), "Free functions cannot be virtual.");
+		else if (_function.annotation().contract->isInterface())
```
This check and a few of the others in the vicinity actually belong in the SyntaxChecker, don't they? But can be done in a separate PR.

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -1820,6 +1846,8 @@ void TypeChecker::typeCheckFunctionCall(
 	if (_functionType->kind() == FunctionType::Kind::Declaration)
 	{
 		if (
+			m_currentContract &&
```
Could this be an assertion instead?

chriseth replied:
No, and this is tested in test/libsolidity/syntaxTests/freeFunctions/free_call_via_contract_type.sol:
```
contract C {
    function f() public pure {}
}
function fun() {
    C.f();
}
// ----
// TypeError 3419: (68-73): Cannot call function via contract type name.
```

ekpyron replied:
Ah, right - I actually meant the other one, one line below (as in: only functions within contract scopes should ever be available with type ``FunctionType::Kind::Declaration``, so ``_functionType->declaration().annotation().contract`` should never be ``nullptr`` here - that's why it could just be dereferenced before this PR, right?).
But also fine not to assert that.

chriseth replied:
Ah this might be true! And it's important to realize this: References to free function can never be `::Declaration`. I'll add a test where we reference a free function via a module name to check if this is really the case.

chriseth replied:
Ah, there already is a test and it does not violate the assertion: https://github.com/ethereum/solidity/pull/8834/commits/756aae0cccfae951e95db62d735c50ed9685116b#diff-9177bb8d3bc85509fb8289a4da7f6085

leonardoalt replied:
> And it's important to realize this: References to free function can never be `::Declaration`.

Why is that?

leonardoalt replied:
because Declaration always needs member access

ekpyron commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -2537,18 +2565,23 @@ void TypeChecker::endVisit(NewExpression const& _newExpression)
-				"Circular reference for contract creation (cannot create instance of derived or same contract)."
+		if (m_currentContract)
+		{
+			// TODO this is not properly detecting creation-cycles if they go through
+			// internal library functions or free functions. It will be caught at
+			// code generation time, but it would of course be better to catch it here.
```
This seems like something worth still doing in this PR... This means ``contractDependencies`` may be incorrect in the presence of free functions? We do export that in the JSON AST, so maybe better not just ignore it...
(Haven't really thought this through, though, so I may be wrong - but seems to me that we need to track contract dependencies for free functions and then propagate them to contracts calling them...)

chriseth replied:
It is already incorrect when using internal library functions, so this PR does not make anything worse.

ekpyron replied:
Ah ok... do we have an issue for it then at least?

chriseth replied:
Created it: https://github.com/ethereum/solidity/issues/9565

ekpyron commented on libsolidity/analysis/ViewPureChecker.h:
```diff
@@ -75,6 +75,7 @@ class ViewPureChecker: private ASTConstVisitor
 	langutil::ErrorReporter& m_errorReporter;
 
 	bool m_errors = false;
+	bool m_checkModifiers = false;
```
I haven't thought this through entirely - but would it be possible to only visit functions and modifiers once, but to visit the referenced modifier when encountering the modifier invocation on a function unless ``m_inferredMutability`` is already populated for that modifier?
To me that would seem cleaner than this two-step approach.

chriseth replied:
it seems to work!

ekpyron reviewed (commented):
Some comments - but generally this looks good.

chriseth committed:
Use opportunistic visiting order for modifiers.

chriseth commented:
The last commit should be squashed.

ekpyron commented on docs/structure-of-a-contract.rst:
```diff
@@ -43,19 +43,26 @@ visibility.
 
     // SPDX-License-Identifier: GPL-3.0
-    pragma solidity >=0.4.0 <0.8.0;
+    pragma solidity >=0.7.1 <0.8.0;
```
```suggestion
    pragma solidity >0.7.0 <0.8.0;
```

The command line tests are failing because of this, right? Or maybe because of something else.

ekpyron commented on libsolidity/ast/AST.h:
```diff
@@ -849,6 +853,7 @@ class FunctionDefinition: public CallableDeclaration, public StructurallyDocumen
 
 private:
 	StateMutability m_stateMutability;
+	bool m_free;
```
Just as a random note without necessarily suggesting to do this in this PR, but if we went ahead with a suggestion like https://github.com/ethereum/solidity/pull/9364#discussion_r459380693 , i.e. assign scopes during parsing already, then we wouldn't need this member, but could derive it from the scope instead...

axic commented on test/cmdlineTests/recovery_ast_constructor/output:
```diff
@@ -58,6 +58,7 @@ JSON AST:
+            "freeFunction": false,
             "implemented": true,
             "isConstructor": true,
             "kind": "constructor",
```
Why do we have `isConstructor` if we have kind? And why introduce a new attribute `freeFunction` instead of extending `kind`?

chriseth replied:
I didn't see "being free" and the "kind of function" to be the same thing, but you are right that they currently all exclude each other.

axic commented on libsolidity/ast/ASTJsonConverter.cpp:
```diff
@@ -467,7 +468,6 @@ bool ASTJsonConverter::visit(UserDefinedTypeName const& _node)
 	setJsonNode(_node, "UserDefinedTypeName", {
 		make_pair("name", namePathToString(_node.namePath())),
 		make_pair("referencedDeclaration", idOrNull(_node.annotation().referencedDeclaration)),
-		make_pair("contractScope", idOrNull(_node.annotation().contractScope)),
```
Why is this removed? Is this a breaking change?

chriseth replied:
It is always null because `contractScope` is never assigned.

axic commented on libsolidity/ast/AST.h:
```diff
@@ -803,6 +805,7 @@ class FunctionDefinition: public CallableDeclaration, public StructurallyDocumen
 	bool isConstructor() const { return m_kind == Token::Constructor; }
 	bool isFallback() const { return m_kind == Token::Fallback; }
 	bool isReceive() const { return m_kind == Token::Receive; }
+	bool isFree() const { return m_free; }
```
Btw where is this "free functions" term coming from? C++?

I think `isFree` looks kind of weird to be honest for anyone not 100% familiar with this (but `isReceive` is a bit similar in that sense).

chriseth replied:
I would call it `free`, but you cannot do that in C++. I'm happy to use a totally different term.

chriseth replied:
I think it is a C++ term, yes: https://en.wikipedia.org/wiki/Free_function

axic commented on libsolidity/analysis/SyntaxChecker.cpp:
```diff
@@ -309,19 +309,41 @@ bool SyntaxChecker::visit(ContractDefinition const& _contract)
+			m_errorReporter.syntaxError(
+				4126_error,
+				_function.location(),
+				"Free functions cannot have visibility."
```
```suggestion
				"Free functions cannot have their visibility specified."
```

axic commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -336,12 +340,16 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 			m_errorReporter.typeError(7708_error, _function.location(), "Library functions cannot be payable.");
-		if (_function.isOrdinary() && !_function.isPartOfExternalInterface())
+		else if (_function.isFree())
+			m_errorReporter.typeError(9559_error, _function.location(), "Free functions cannot be payable.");
```
If we are placing so many restrictions, I think the "free functions" term is pretty much a misnomer, and based on the documentation piece, these would be just called "inline functions".

chriseth replied:
internal?

leonardoalt replied:
I guess they are kind of library functions but free from libraries

chriseth committed:
Use kind in json AST for free function.

chriseth commented on libsolidity/ast/ASTJsonConverter.cpp:
```diff
@@ -360,7 +360,7 @@ bool ASTJsonConverter::visit(FunctionDefinition const& _node)
 		make_pair("name", _node.name()),
 		make_pair("documentation", _node.documentation() ? toJson(*_node.documentation()) : Json::nullValue),
-		make_pair("kind", TokenTraits::toString(_node.kind())),
+		make_pair("kind", _node.isFree() ? "freeFunction" : TokenTraits::toString(_node.kind())),
```
Should we just use "free"?

chriseth replied:
I'll change it to "free".

chriseth replied:
Actually no, I think it's fine.

chriseth commented:
Squashed the "to be squashed" commit.

leonardoalt commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -95,6 +95,7 @@ bool TypeChecker::visit(ContractDefinition const& _contract)
 
 	m_currentContract = nullptr;
 
+	m_currentContract = nullptr;
```
Duplicate, two lines above.

leonardoalt commented on libsolidity/analysis/ViewPureChecker.cpp:
```diff
@@ -128,28 +128,12 @@ class AssemblyViewPureChecker
-			mod->accept(*this);
-
-	for (auto const& contract: contracts)
-		contract->accept(*this);
```
Where did this block go? Is there a visitor for SourceUnit?

chriseth replied:
SourceUnit is the root AST node and the default visitor just descends into its children

leonardoalt commented on test/libsolidity/syntaxTests/freeFunctions/free_call_via_contract_type.sol:
```diff
@@ -0,0 +1,8 @@
+    C.f();
+}
+// ----
+// TypeError 3419: (68-73): Cannot call function via contract type name.
```
If it's pure, why not?

chriseth replied:
You still need an instance. But you are right, most pure functions can probably be turned into free functions which then can be called without an instance.

leonardoalt commented on test/libsolidity/syntaxTests/freeFunctions/free_function_visibility.sol:
```diff
@@ -0,0 +1,4 @@
+function fun() public {
+}
+// ----
+// SyntaxError 4126: (0-25): Free functions cannot have visibility.
```
Does it say the same if you give `internal`?

chriseth replied:
Yes, will add another function with internal visibility.

leonardoalt commented on test/libsolidity/syntaxTests/freeFunctions/free_functions.sol:
```diff
@@ -0,0 +1,4 @@
+function fun(uint256, uint[] calldata _x, uint[] storage _y) view  returns (uint, uint[] calldata) {
+	return (_y[0], _x);
+}
+// ----
```
How about a semantic test like this?

leonardoalt reviewed (changes requested):
Looks good overall, left some questions.

Do you think it makes sense to add semantic tests for:
- Storage/memory references
- Calls to a contract
- Recursive free functions

chriseth commented:
The `new_operator` test already calls a function on a contract. Will add recursion test and the calldata / storage reference test.

chriseth commented:
Updated.

leonardoalt commented:
There are a few items from the `test ideas` (first comment) that are not in the tests. Do you still want to add those, or fine without?

chriseth commented:
The thing is that free functions and internal library functions are really the same thing, so I would say we are pretty safe not adding so many tests. I will add another test that uses free functions via imports.

leonardoalt commented:
Yea, that's also what I thought. There is already a test that uses imports though
